src/Engine/EmitMSIL/CodeGenIL.cs

```diff
@@ -115,16 +113,182 @@ public CodeGenIL(IDictionary<string, IList> input, string filePath)
 
         private Type EmitCoercionCode(Type argType, ParameterInfo param)
         {
+            if (param.ParameterType.IsAssignableFrom(argType)) return argType;
+
             if(argType == typeof(double) && param.ParameterType == typeof(long))
             {
                 EmitOpCode(OpCodes.Conv_I8);
                 return typeof(long);
             }
+            if (argType == typeof(double) || argType == typeof(long) && param.ParameterType == typeof(int))
+            {
+                EmitOpCode(OpCodes.Conv_I4);
+                return typeof(int);
+            }
+            if(argType == typeof(double[]) && typeof(IEnumerable<int>).IsAssignableFrom(param.ParameterType))
+            {
+                return EmitArrayCoercion<double, int>();
+            }
+            if(argType == typeof(int[]) && typeof(IEnumerable<double>).IsAssignableFrom(param.ParameterType))
+            {
+                return EmitArrayCoercion<int, double>();
+            }
+            if (argType == typeof(double[]) && typeof(IEnumerable<long>).IsAssignableFrom(param.ParameterType))
+            {
+                return EmitArrayCoercion<double, long>();
+            }
+            if (argType == typeof(long[]) && typeof(IEnumerable<double>).IsAssignableFrom(param.ParameterType))
+            {
+                return EmitArrayCoercion<long, double>();
+            }
+            // TODO: The following cases do not work yet as EmitIListCoercion needs to be fixed.
+            //if (typeof(IList).IsAssignableFrom(argType) && typeof(IEnumerable<double>).IsAssignableFrom(param.ParameterType))
+            //{
+            //    return EmitIListCoercion<double>();
+            //}
+            //if (typeof(IList).IsAssignableFrom(argType)  && typeof(IEnumerable<int>).IsAssignableFrom(param.ParameterType))
+            //{
+            //    return EmitIListCoercion<int>();
+            //}
+            //if (typeof(IList).IsAssignableFrom(argType) && typeof(IEnumerable<long>).IsAssignableFrom(param.ParameterType))
+            //{
+            //    return EmitIListCoercion<long>();
+            //}
             // TODO: Add more coercion cases here.
 
             return argType;
         }
 
+        // TODO: Does not work yet (intended to coerce IList to IEnumerable<T>)
+        private Type EmitIListCoercion<T>()
+        {
+            MethodInfo mInfo = null;
+            // Find length for IList to be coerced (already on top of eval stack), len
+            var prop = typeof(ICollection).GetProperties(BindingFlags.Public | BindingFlags.Instance).Where(
+                                p => p.Name == nameof(ICollection.Count)).FirstOrDefault();
+
+            if (prop != null)
+            {
+                mInfo = prop.GetAccessors().FirstOrDefault();
+            }
```

Comment:
if this is null, shouldn't we throw an exception?

Reply:
I guess so. Will add it.